presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java

```diff
@@ -194,4 +218,22 @@ public Symbol getRight()
             return right;
         }
     }
+
+    public boolean isDistributed()
+    {
+        return method == DISTRIBUTED;
+    }
+
+    public static boolean canDistributeJoin(Session session, PlanNode rightNode, List<JoinNode.EquiJoinClause> equiJoinClauses, JoinNode.Type type)
```

Comment:
I do not like calling this `must/canDistributeJoin` in many optimizers/planner. This is error prone and exposes current limitations of execution engine to optmizers. I would like to avoid that.

Maybe instead optimizers/planner should just assume that both methods `BROADCAST` and `DISTRIBUTED` is always possible for each `JoinNode`. And just plan according to that assumption.

And then add another optimizer at the end which will change method from `BROADCAST` to `DISTRIBUTED` for joins which cannot execute the other way around. Or even to all we enforce that all joins should be DISTRIBUTED via session property.